Documentation/proposals/202212-scrape-config.md

```diff
@@ -0,0 +1,99 @@
+# ScrapeConfig CRD
+* Owners:
+  * [xiu](https://github.com/xiu)
+* Related Tickets:
+  * [#2787](https://github.com/prometheus-operator/prometheus-operator/issues/2787)
+  * [#3447](https://github.com/prometheus-operator/prometheus-operator/issues/3447)
+* Other docs:
+  * n/a
+
+This document aims at creating a lower level `ScrapeConfig` Custom Resource that defines additional scrape configurations the kubernetes way.
+
+# Why
+
+prometheus-operator misses a way to scrape external targets using CRD. Users have either been abusing the Probe CRD (#3447) or additionalScrapeConfig to do so.
+Furthermore, currently, there is a lot of code duplication due to the operator supporting several CRDs that generate scrape configurations. With the new `ScrapeConfig` CRD, it would be possible to consolidate some of that logic, where the other `*Monitor` CRDs could be migrated so that they create a ScrapeConfig resource that would ultimately be used by the operator to generate scrape configuration.
+
+## Pitfalls of the current solution
+
+Using `additionalScrapeConfig` comes with drawbacks:
+* it becomes a bottleneck as teams have to build an infrastructure to add scrape rules in a centralized manner, which creates a bottleneck since a team becomes responsible for the configuration
+* there is no input validation, which can lead to an invalid prometheus configuration
+
+# Goals
+* Provide a way for users to self-service adding scrape targets
+* Consolidate the scrape configuration generation logic in a central point for other resources to use
+
+## Audience
+* Users who serve prometheus as a service and want to have their customers autonomous in defining scrape configs
+* Users who want to manage scrape configs the same way as for services runnings within the kubernetes cluster
+
+# Non-Goals
+* this proposal doesn't aim at covering all the fields in [`<scrape_config>`](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config). Specifically, no service discovery other than `static_configs` or `file_sd_configs` should be implemented at first.
+* refactoring of the other CRDs is not in scope for the first version
+
+# How
+
+As described by [@aulig in #2787](https://github.com/prometheus-operator/prometheus-operator/issues/2787#issuecomment-559776221), we will create a new `ScrapeConfig` CRD, this config will act the same as the other CRDs and append scrape configurations to the configuration. `ScrapeConfig` will allow for any scraping configuration, while the other CRDs provide sane defaults. This will allow for isolated testing of the new `ScrapeConfig` CRD.
+
+```mermaid
+graph TD;
+    ServiceMonitor-->prometheusConfig
+    PodMonitor-->prometheusConfig
+    ScrapeConfig-->prometheusConfig
+    prometheusConfig
+```
+
+Using a pseudo custom resource definition, we should have the following:
+
+```yaml
```

Comment:
I was more thinking about it in golang 😅 but up to you

```golang
// Scrape Config defines a prometheus scrape_config.
type ScrapeConfig struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`
	// Specification of the desired scrape configuration More info:
	// https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
	Spec ScrapeConfigSpec `json:"spec"`
}

// ScrapeConfigSpec is a specification of the desired behavior of the ScrapeConfig. More info:
// https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
// +k8s:openapi-gen=true
type ScrapeConfigSpec struct {
         // The list of static configs
        StaticConfigs  []StaticConfig `json:"staticConfigs,omitempty"`
         // The list of file service discovery configs
        FileSDConfigs []SecretOrConfigMap `json:"fileSDConfig,omitempty"`
        // The list of relabel configurations.
        RelabelConfigs []RelableConfig `json:"reliableConfig,omitempty"`
}

// +k8s:openapi-gen=true
type StaticConfig {
         Targets []string `json:"targets"`
         Labels map[string]string `json:"labels,omitempty"`
}

// Already exists
type SecretOrConfigMap 

// Already exists
type RelabelConfigs
```

I'm not sure if I'm missing something

Reply:
Right, I didn't want to be too strict in the definition as that definition might change as we go about implementing it. @simonpasquier mentioned something about not having the definition in the proposal in the contributor office hours.

Reply:
I'm fine with a raw YAML version here. The details can be discussed in pull requests.